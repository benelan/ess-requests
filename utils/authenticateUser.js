/**
 * Authentication workflows using ArcGIS OAuth
 *  @module authenticateUser
 * */
import { loadModules } from 'esri-loader'

const APP_ID = 'hZpRbKz3fiSkaj1U'

/**
   * ArcGIS OAuth login
   * @async
   * @func
   * @return {object} user's name and email
   */
export async function esriLogin() {
  const [Portal, OAuthInfo, IdentityManager] = await (loadModules([
    'esri/portal/Portal',
    'esri/identity/OAuthInfo',
    'esri/identity/IdentityManager',
  ]))

  const info = new OAuthInfo({
    appId: APP_ID,
    popup: false,
  })

  IdentityManager.registerOAuthInfos([info])
  IdentityManager.getCredential(`${info.portalUrl}/sharing`)
  return new Promise((resolve, reject) => {
    IdentityManager.checkSignInStatus(`${info.portalUrl}/sharing`)
      .then(() => {
        const portal = new Portal()
        // Setting authMode to immediate signs the user in once loaded
        portal.authMode = 'immediate'
        // Once loaded, user is signed in
        portal.load().then(() => {
          // set store values of email and name
          resolve({ name: portal.user.fullName, email: portal.user.email })
        })
      })
      .catch((e) => {
        reject(e)
      })
  })
}

/**
   * ArcGIS OAuth logout
   * @async
   * @func
   */
export async function esriLogout() {
  // load modules
  const [OAuthInfo, IdentityManager] = await (loadModules([
    'esri/identity/OAuthInfo',
    'esri/identity/IdentityManager',
  ]))

  // destroy credentials
  const info = new OAuthInfo({
    appId: APP_ID,
    popup: false,
  })
  IdentityManager.registerOAuthInfos([info])
  IdentityManager.destroyCredentials()
}
