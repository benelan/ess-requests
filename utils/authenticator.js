import { loadModules } from 'esri-loader'

export async function esriLogin() {
  const [Portal, OAuthInfo, IdentityManager] = await (loadModules([
    "esri/portal/Portal",
    "esri/identity/OAuthInfo",
    "esri/identity/IdentityManager"
  ]))

  var info = new OAuthInfo({
    appId: "hZpRbKz3fiSkaj1U",
    popup: false
  })

  IdentityManager.registerOAuthInfos([info])
  IdentityManager.getCredential(info.portalUrl + "/sharing")
  return new Promise((resolve, reject) => {
    IdentityManager.checkSignInStatus(info.portalUrl + "/sharing")
      .then(() => {
        var portal = new Portal()
        // Setting authMode to immediate signs the user in once loaded
        portal.authMode = "immediate"
        // Once loaded, user is signed in
        portal.load().then(function () {
          // set store values of email and name
          resolve({name: portal.user.fullName, email: portal.user.email})
        })
      })
      .catch((e) => {
        reject(e)
      })
  })
}

export async function esriLogout() {
  // load modules
  const [OAuthInfo, IdentityManager] = await (loadModules([
    "esri/identity/OAuthInfo",
    "esri/identity/IdentityManager"
  ]))

  // destroy credentials
  var info = new OAuthInfo({
    appId: "hZpRbKz3fiSkaj1U",
    popup: false
  })
  IdentityManager.registerOAuthInfos([info])
  IdentityManager.destroyCredentials()
}