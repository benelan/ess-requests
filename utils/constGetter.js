/** @module constGetter */

/**
   * @func
   * @return {array} employee locations
   */
export const getEmployeeLocations = () => ['Redlands', 'Charlotte', 'Washington DC', 'St Louis']

/**
   * @func
   * @return {array} exam locations
   */
export const getExamLocations = () => ['Redlands', 'Charlotte', 'Washington DC', 'St Louis', 'Remote']

/**
   * @func
   * @return {object} employee units
   */
export const getUnits = () => ({
  Online: 'Supt-ArcGIS-Unit-Mgmt@esri.com',
  Enterprise: 'Supt-Enterprise-Unit-Mgmt@esri.com',
  Desktop: 'Supt-Desktop-Unit-Mgmt@esri.com',
  'DaDP Product': 'Supt-DaDP-Unit-Mgmt@esri.co',
  NORUS: 'Supt-NORUS-Unit-Mgmt@esri.com',
  TAMS: 'SWhittington@esri.com',
  Readiness: 'Christian_Wells@esri.com',
})

/**
   * @func
   * @param {string} unit
   * @param {string} location
   * @return {string} cost center
   */
export const getCostCenter = (unit, location) => {
  if (unit === 'Supt-NORUS-Unit-Mgmt@esri.com') {
    return '4255'
  }
  switch (location) {
  case 'Redlands':
    return '4252'
  case 'Charlotte':
    return '4253'
  case 'St Louis':
    return '4252'
  default:
    return '4255'
  }
}

/**
   * @func
   * @param {string} unit
   * @return {string} charge code
   */
export const getChargeCode = (unit) => {
  switch (unit) {
  case 'Supt-NORUS-Unit-Mgmt@esri.com':
    return 'TE0702'
  default:
    return 'TE0352'
  }
}
