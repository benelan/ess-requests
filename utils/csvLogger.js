/** @module csvLogger */
import fs from 'fs'
import jsonexport from 'jsonexport'

/**
   * receives JSON data from POST and converts it to JSON
   * Either creates a new csv file or appends data to an existing file
   * @func
   * @param {object} res - REST response
   * @param {object} data - JSON object with form data
   * @param {string} filePath - the path to the csv file
   * @return {promise}
 */
export default async (res, data, filePath) => {
  return new Promise(resolve => {
    // handles successful response
    function successful() {
      res.statusCode = 200
      res.json({ response: 'logging successful' })
      return resolve()
    }

    // handles failed response
    function fail(failureError) {
      console.error(failureError)
      res.statusCode = 500
      res.json({ response: failureError })
      return resolve()
    }

    // convert the request data to csv format
    jsonexport([JSON.parse(data)], function (exportError, csv) {
      if (exportError) fail(exportError)

      // check if there is a csv file
      fs.stat(filePath, function (noFile) {
        if (noFile) {  // if there is not a csv file create one and write the headers and data
          fs.writeFile(filePath, csv + "\r\n", (createFail) => createFail ? fail(createFail) : successful())
        }
        else {  // if there is a csv file remove the header and append the data
          const noHeadersCSV = csv.split('\n').pop() + "\r\n"
          fs.appendFile(filePath, noHeadersCSV, (appendFail) => appendFail ? fail(appendFail) : successful())
        }
      })
    })
  })
}