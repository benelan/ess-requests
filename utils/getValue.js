/**
 * Named function exports that return hardcoded values
 * @module getValue
 * */

/**
 * The location of ESS offices
 * @func
 * @return {array} ESS office locations
 */
export const getOfficeLocations = () => ['Redlands', 'Charlotte', 'Washington DC', 'St Louis']

/**
 * The email addresses for requests by unit
 * @func
 * @return {object} employee units
 */
export const getUnits = () => ({
  Online: 'online@fake.com',
  Enterprise: 'enterprise@fake.com',
  Desktop: 'desktop@fake.com',
  'DaD Products': 'dadp@fake.com',
  NORUS: 'norus@fake.com',
  TAMS: 'tams@fake.com',
  Readiness: 'readiness@fake.com',
})

/**
 * Determines cost center by employee unit and location
 * @func
 * @param {string} unit
 * @param {string} location
 * @return {string} cost center
 */
export const getCostCenter = (unit, location) => {
  if (unit === 'norus@fake.com') {
    return '4255'
  }
  switch (location) {
  case 'Redlands':
    return '4252'
  case 'Charlotte':
    return '4253'
  case 'St Louis':
    return '4252'
  default:
    return '4255'
  }
}

/**
 * Determines charge code by employee unit
 * @func
 * @param {string} unit
 * @return {string} charge code
 */
export const getChargeCode = (unit) => {
  switch (unit) {
  case 'norus@fake.com':
    return 'TE0702'
  default:
    return 'TE0352'
  }
}
