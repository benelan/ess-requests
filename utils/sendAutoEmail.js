/* eslint-disable import/prefer-default-export */
/**
 * Logs form data to csv
 * @module sendAutoEmail
 * */
import axios from 'axios'
import { flowUrl } from '../email.config'

/**
 * Automatically sends an email request
 * @param {object} { subject, text, to } - the email information
 * @return {boolean} did the email send successfully
 */
export const sendAutoEmail = async ({ subject, text, to }) => new Promise((resolve, reject) => {
  axios.post(
    flowUrl,
    { to, subject, text: text.replaceAll(',', '<br>') },
    {
      headers: { 'Content-Type': 'application/json' },
    },
  )
    .then((response) => resolve(response.data === 'sent'))
    .catch((error) => reject(error))
})
