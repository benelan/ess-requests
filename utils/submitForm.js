/**
 * This module handles form submission workflows. The exported named function submitForm determines if the data is valid and which type of form to submit. It then sends a POST to the API to log the request to CSV and send an email. If the email does not send on the server, a template will be generated on the client.
 * @module submitForm
 * */

import { examSchemaIsValid, trainingSchemaIsValid } from './validateData'

/**
   * Generates and returns an email with the training request data
   * @func
   * @param {object} formData - the form data and costCenter and chargeCode
   * @return {string} mailto string
 */
export const generateTrainingEmail = (formData) => {
  try {
    const to = `${formData.unit}, ${formData.emailEmployee}`
    const subject = 'Request for Training'
    const text = `Employee Name: ${formData.nameEmployee}
Employee Email: ${formData.emailEmployee}
Employee Number: ${formData.numberEmployee}
Cost Center: ${formData.costCenter}
Employee Location: ${formData.locationEmployee}
Charge Code: ${formData.chargeCode}
Course Name: ${formData.nameCourse}
Cost: ${formData.cost}
Start Date: ${formData.startDate}
End Date: ${formData.endDate}
Vendor: ${formData.vendor}
Justification: ${formData.justification}
Comments: ${formData.comments}`

    return { subject, text, to }
  } catch (e) {
    console.error(e)
    return null
  }
}

/**
   * Generates and returns an email with the exam request data
   * @func
   * @param {object} formData - the form data and costCenter and chargeCode
   * @return {string} mailto string
 */
export const generateExamEmail = (formData) => {
  try {
    const to = `${formData.unit}, ${formData.emailEmployee}`
    const subject = 'Request for Exam Certification'
    const text = `Employee Name: ${formData.nameEmployee}
Employee Email: ${formData.emailEmployee}
Employee Number: ${formData.numberEmployee}
Cost Center: ${formData.costCenter}
Employee Location: ${formData.locationEmployee}
Charge Code: ${formData.chargeCode}
Exam Name: ${formData.nameExam}
Exam Cost: $${formData.cost}
Exam Testing Location: ${formData.locationExam}
Exam Vendor: ${formData.vendor}
Justification: ${formData.justification}`

    return { subject, text, to }
  } catch (e) {
    console.error(e)
    return null
  }
}

/**
   * Sends a POST to the api to log the training request form data to csv
   * @func
   * @param {object} formData - the form data and costCenter and chargeCode
 */
const apiTrainingPOST = async (formData) => new Promise((resolve, reject) => {
  fetch(`${process.env.basePath}/api/submitTrainingRequest`, {
    method: 'post',
    body: JSON.stringify(formData),
  })
    .then((response) => response.json())
    .then((data) => resolve(data))
    .catch((e) => reject(e))
})

/**
   * Sends a POST to the api to log the exam request form data to csv
   * @func
   * @param {object} formData - the form data and costCenter and chargeCode
 */
const apiExamPOST = async (formData) => new Promise((resolve, reject) => {
  fetch(`${process.env.basePath}/api/submitExamRequest`, {
    method: 'post',
    body: JSON.stringify(formData),
  })
    .then((response) => response.json())
    .then((data) => resolve(data))
    .catch((e) => reject(e))
})

/**
     * Validates data by type and schema, then submits the form
     * @func
     * @param {string} formType - the type of form being submitted
     * @param {string} inputData - the form data
     * @return {string} mailto string
   */
export const submitForm = async (formType, inputData) => {
  try {
    // check training form schema validity
    if (formType === 'training' && trainingSchemaIsValid(inputData)) {
      // send data to server for csv logging and emailing
      const response = await apiTrainingPOST(inputData)
      // if the server was unable to send the email
      if (!response.sent) {
        // generate the email on the client side
        const { text, subject } = generateTrainingEmail(inputData)
        // create and return a mailto string
        return `mailto:${inputData.unit}?subject=${encodeURIComponent(subject)}&body=${encodeURIComponent(text)}`
      }
      // if the email sent from the server no need to return mailto string
      return 'sent'
    }
    // same steps as above, only for exam forms
    if (formType === 'exam' && examSchemaIsValid(inputData)) {
      const response = await apiExamPOST(inputData)
      if (!response.sent) {
        const { text, subject } = generateExamEmail(inputData)

        return `mailto:${inputData.unit}?subject=${encodeURIComponent(subject)}&body=${encodeURIComponent(text)}`
      }
      return 'sent'
    }
    // if there were issues with the schemas
    console.error('invalid data schema')
    return 'error'
  } catch (e) {
    console.error(e)
    return 'error'
  }
}
