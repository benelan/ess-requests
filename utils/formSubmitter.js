/* eslint-disable no-alert */
/* eslint-disable no-restricted-globals */
/** @module formSubmitter */
import getConfig from 'next/config'
import { getChargeCode, getCostCenter } from './constGetter'

const { publicRuntimeConfig } = getConfig()

/**
   * Generates and returns an email with the training request data
   * @func
   * @param {object} formData - the form data and costCenter and chargeCode
   * @return {string} mailto string
 */
export const generateTrainingEmail = (formData) => {
  try {
    const subject = 'Request for Training'
    const body = `Employee Name: ${formData.nameEmployee}
Employee Email: ${formData.emailEmployee}
Employee Number: ${formData.numberEmployee}
Cost Center: ${formData.costCenter}
Employee Location: ${formData.locationEmployee}
Charge Code: ${formData.chargeCode}
Course Name: ${formData.nameCourse}
Cost: ${formData.cost}
Start Date: ${formData.startDate}
End Date: ${formData.endDate}
Vendor: ${formData.vendor}
Justification: ${formData.justification}
Comments: ${formData.comments}`

    return `mailto:${formData.unit}?cc=${formData.emailEmployee}&subject=${encodeURIComponent(subject)}&body=${encodeURIComponent(body)}`
  } catch (e) {
    console.error(e)
    return null
  }
}

/**
   * Sends a POST to the api to log the training request form data to csv
   * @func
   * @param {object} formData - the form data and costCenter and chargeCode
 */
export const logTrainingRequest = (formData) => {
  try {
    const outputData = {
      'Employee Name': formData.nameEmployee,
      'Employee Email': formData.emailEmployee,
      'Employee Number': formData.numberEmployee,
      'Employee Location': formData.locationEmployee,
      'Cost Center': formData.costCenter,
      'Charge Code': formData.chargeCode,
      'Course Name': formData.nameCourse,
      'Exam Cost': formData.cost,
      'Exam Vendor': formData.vendor,
      'Start Date': formData.startDate,
      'End Date': formData.endDate,
      Comments: formData.comments,
      Justification: formData.justification,
    }

    fetch(`${publicRuntimeConfig.basePath}/api/logTraining`, {
      method: 'post',
      body: JSON.stringify(outputData),
    }).then((response) => response.json()).then((data) => {
      console.log(data.message)
    })
  } catch (e) {
    console.error(e)
  }
}

/**
   * Generates and returns an email with the exam request data
   * @func
   * @param {object} formData - the form data and costCenter and chargeCode
   * @return {string} mailto string
 */
export const generateExamEmail = (formData) => {
  try {
    const subject = 'Request for Exam Certification'
    const body = `Employee Name: ${formData.nameEmployee}
Employee Email: ${formData.emailEmployee}
Employee Number: ${formData.numberEmployee}
Cost Center: ${formData.costCenter}
Employee Location: ${formData.locationEmployee}
Charge Code: ${formData.chargeCode}
Exam Name: ${formData.nameExam}
Exam Cost: $${formData.cost}
Exam Testing Location: ${formData.locationExam}
Exam Vendor: ${formData.vendor}
Justification: ${formData.justification}`

    return `mailto:${formData.unit}?cc=${formData.emailEmployee}&subject=${encodeURIComponent(subject)}&body=${encodeURIComponent(body)}`
  } catch (e) {
    console.error(e)
    return null
  }
}

/**
   * Sends a POST to the api to log the exam request form data to csv
   * @func
   * @param {object} formData - the form data and costCenter and chargeCode
 */
export const logExamRequest = (formData) => {
  try {
    const outputData = {
      'Employee Name': formData.nameEmployee,
      'Employee Email': formData.emailEmployee,
      'Employee Number': formData.numberEmployee,
      'Employee Location': formData.locationEmployee,
      'Cost Center': formData.costCenter,
      'Charge Code': formData.chargeCode,
      'Exam Name': formData.nameExam,
      'Exam Cost': formData.cost,
      'Exam Testing Location': formData.locationExam,
      'Exam Vendor': formData.vendor,
      Justification: formData.justification,
    }

    fetch(`${publicRuntimeConfig.basePath}/api/logExam`, {
      method: 'post',
      body: JSON.stringify(outputData),
    }).then((response) => response.json()).then((data) => {
      console.log(data.message)
    })
  } catch (e) {
    console.error(e)
  }
}

/**
     * Handles valid submits based on form type
     * @func
     * @param {string} type - the type of form being submitted
     * @param {string} state - the form data
   */
export const handleValidSubmit = (type, state) => {
  try {
    const chargeCode = getChargeCode(state.unit)
    // have user confirm submission and inform them of the charge code
    const confirmed = confirm(`Please remember your charge code: ${chargeCode}`)
    if (confirmed) {
      const costCenter = getCostCenter(state.unit, state.locationEmployee)
      const formData = { ...state, costCenter, chargeCode }
      let mailtoString = ''
      if (type === 'training') {
        logTrainingRequest(formData)
        mailtoString = generateTrainingEmail(formData)
      }
      if (type === 'exam') {
        logExamRequest(formData)
        mailtoString = generateExamEmail(formData)
      }
      if (mailtoString) window.open(mailtoString)
    }
  } catch (e) {
    console.error(e)
  }
}
