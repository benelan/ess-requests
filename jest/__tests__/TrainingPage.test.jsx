import React from 'react'
import {
  render, screen, fireEvent, waitFor,
} from '@testing-library/react'
import renderer from 'react-test-renderer'
import TrainingPage from '../../pages/training'
import * as submit from '../../utils/submitForm'

const { act } = renderer

describe('<TrainingPage /> snapshot', () => {
  it('matches component', () => {
    const examTree = renderer.create(<TrainingPage />).toJSON()
    expect(examTree).toMatchSnapshot()
  })
})

describe('<TrainingPage /> renders', () => {
  it('user name', async () => {
    render(<TrainingPage />)
    expect(
      await screen.findByRole('textbox', { name: 'Employee Name' }),
    ).toHaveValue('Test User')
  })

  it('navbar with training active', async () => {
    render(<TrainingPage />)
    await act(async () => screen.findByRole('textbox', { name: 'Employee Name' }))
    expect(screen.getByRole('listitem', { name: 'Training Link' })).toHaveClass('active')
  })
})

describe('<TrainingPage /> events', () => {
  it('doesn\'t submit empty form', async () => {
    render(<TrainingPage />)
    await act(async () => screen.findByRole('textbox', { name: 'Employee Name' }))
    fireEvent.click(screen.getByRole('button', { name: 'Submit' }))

    // bootstrap validation
    expect(
      screen.getByRole('form', { name: /Training Request Form/ }),
    ).toHaveClass('was-validated')

    // submit doesn't happen
    expect(window.confirm).not.toBeCalled()
    expect(window.fetch).not.toBeCalled()
  })

  it('submits filled in form', async () => {
    render(<TrainingPage />)
    // wait for user name to populate after mocked auth
    await act(async () => screen.findByRole('textbox', { name: 'Employee Name' }))

    // fill in the form
    fireEvent.change(
      screen.getByRole('spinbutton', { name: 'Employee Number' }),
      { target: { value: '12345' } },
    )
    fireEvent.change(screen.getByRole('spinbutton', { name: 'Cost' }), {
      target: { value: '54321' },
    })
    fireEvent.change(
      screen.getByRole('combobox', { name: 'Employee Location' }),
      { target: { value: 'St Louis' } },
    )
    fireEvent.change(screen.getByRole('combobox', { name: 'Employee Unit' }), {
      target: { value: 'Enterprise' },
    })
    fireEvent.change(screen.getByRole('textbox', { name: 'Course Name' }), {
      target: { value: 'Test Course' },
    })
    fireEvent.change(screen.getByRole('textbox', { name: 'Vendor' }), {
      target: { value: 'GIS for Dummies' },
    })
    fireEvent.change(screen.getByRole('textbox', { name: 'Justification' }), {
      target: { value: 'I want to be smart' },
    })
    fireEvent.change(screen.getByRole('textbox', { name: 'Comments' }), {
      target: { value: 'Thanks!' },
    })
    fireEvent.change(screen.getByLabelText(/Start Date/), {
      target: { value: '2020-12-22' },
    })
    fireEvent.change(screen.getByLabelText(/End Date/), {
      target: { value: '2020-12-22' },
    })

    // spy on submit function
    const submitMock = jest.spyOn(submit, 'submitForm')

    // submit the form
    act(() => {
      fireEvent.click(screen.getByRole('button', { name: 'Submit' }))
    })
    // form submission succeeds when all inputs are filled out
    await waitFor(() => expect(window.fetch).toReturn())
    // mailto string is only created on error when the email doesn't auto send
    await waitFor(() => expect(submitMock).toReturn())
  })
})
