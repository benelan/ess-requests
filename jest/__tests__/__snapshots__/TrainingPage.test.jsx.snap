// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrainingPage /> snapshot matches component 1`] = `
Array [
  <title>
    Training Request
  </title>,
  <nav
    className="navbar navbar-expand navbar-light border-bottom"
    style={
      Object {
        "background": "white",
      }
    }
  >
    <ul
      className="navbar-nav"
    >
      <li
        aria-label="Home Link"
        className="nav-item"
      >
        <a
          className="nav-link"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Home
        </a>
      </li>
      <li
        aria-label="Training Link"
        className="nav-item active"
      >
        <a
          className="nav-link"
          href="/training"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Training
        </a>
      </li>
      <li
        aria-label="Exam Link"
        className="nav-item"
      >
        <a
          className="nav-link"
          href="/exam"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Exam
        </a>
      </li>
    </ul>
  </nav>,
  <h3
    className="text-center m-4"
  >
    Request for Training
  </h3>,
  <div
    className="container"
  >
    <form
      aria-label="Training Request Form"
      className="needs-validation"
      noValidate={true}
      onSubmit={[Function]}
    >
      <div
        className="form-row mt-3"
      >
        <div
          className="form-group col-md-3"
        >
          <label
            htmlFor="employeeName"
          >
            Employee Name
          </label>
          <input
            className="form-control"
            disabled={true}
            id="employeeName"
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group col-md-3"
        >
          <label
            htmlFor="employeeNumber"
          >
            Employee Number
          </label>
          <input
            className="form-control"
            id="employeeNumber"
            onChange={[Function]}
            required={true}
            type="number"
          />
        </div>
        <div
          className="form-group col-md-3"
        >
          <label
            htmlFor="employeeUnit"
          >
            Employee Unit
          </label>
          <select
            className="form-control"
            defaultValue="prompt"
            id="employeeUnit"
            onChange={[Function]}
          >
            <option
              disabled={true}
              hidden={true}
              value="prompt"
            >
              Select a unit
            </option>
            <option
              value="Supt-ArcGIS-Unit-Mgmt@esri.com"
            >
              Online
            </option>
            <option
              value="Supt-Enterprise-Unit-Mgmt@esri.com"
            >
              Enterprise
            </option>
            <option
              value="Supt-Desktop-Unit-Mgmt@esri.com"
            >
              Desktop
            </option>
            <option
              value="Supt-DaDP-Unit-Mgmt@esri.com"
            >
              DaD Products
            </option>
            <option
              value="Supt-NORUS-Unit-Mgmt@esri.com"
            >
              NORUS
            </option>
            <option
              value="SWhittington@esri.com"
            >
              TAMs
            </option>
            <option
              value="Christian_Wells@esri.com"
            >
              Readiness
            </option>
          </select>
        </div>
        <div
          className="form-group col-md-3"
        >
          <label
            htmlFor="employeeLocation"
          >
            Employee Location
          </label>
          <select
            className="form-control"
            id="employeeLocation"
            onChange={[Function]}
          >
            <option>
              Redlands
            </option>
            <option>
              Charlotte
            </option>
            <option>
              Washington DC
            </option>
            <option>
              St Louis
            </option>
          </select>
        </div>
      </div>
      <div
        className="form-row mt-3"
      >
        <div
          className="form-group col-md-5"
        >
          <label
            htmlFor="courseName"
          >
            Course Name
          </label>
          <input
            className="form-control"
            id="courseName"
            onChange={[Function]}
            required={true}
            type="text"
          />
        </div>
        <div
          className="form-group col-md-5"
        >
          <label
            htmlFor="vendor"
          >
            Vendor
          </label>
          <input
            className="form-control"
            id="vendor"
            onChange={[Function]}
            required={true}
            type="text"
          />
        </div>
        <div
          className="form-group col-md-2"
        >
          <label
            htmlFor="cost"
          >
            Cost
          </label>
          <div
            className="input-group mb-3"
          >
            <div
              className="input-group-prepend"
            >
              <span
                className="input-group-text"
              >
                $
              </span>
            </div>
            <input
              className="form-control"
              id="cost"
              onChange={[Function]}
              required={true}
              type="number"
            />
          </div>
        </div>
      </div>
      <div
        className="form-row"
      >
        <div
          className="form-group col-md-6"
        >
          <label
            htmlFor="justification"
          >
            Justification
          </label>
          <textarea
            className="form-control"
            id="justification"
            onChange={[Function]}
            required={true}
            rows="2"
          />
        </div>
        <div
          className="form-group col-md-6"
        >
          <label
            htmlFor="comments"
          >
            Comments
          </label>
          <textarea
            className="form-control"
            id="comments"
            onChange={[Function]}
            rows="2"
          />
        </div>
      </div>
      <div
        className="form-row mt-3"
      >
        <div
          className="form-group col-md-4"
        >
          <label
            htmlFor="startDate"
          >
            Start Date
          </label>
          <input
            className="form-control"
            id="startDate"
            onChange={[Function]}
            required={true}
            type="date"
          />
        </div>
        <div
          className="form-group col-md-4 offset-md-1"
        >
          <label
            htmlFor="endDate"
          >
            End Date
          </label>
          <input
            className="form-control"
            id="endDate"
            onChange={[Function]}
            required={true}
            type="date"
          />
        </div>
        <div
          className="form-group col-md-2 offset-md-1"
          style={
            Object {
              "marginTop": "31px",
            }
          }
        >
          <input
            className="btn btn-primary"
            disabled={true}
            id="submitButton"
            type="submit"
            value="Submit"
          />
        </div>
      </div>
    </form>
  </div>,
]
`;
